### ISSUE TITLE
<title>Xonsh throws exceptions on initialization only when carapace is added to the run control

### ISSUE THREAD

**Issue by JeffMelton** [carapace-sh/carapace#1068]
### Current Behavior

After adding `exec($(carapace _carapace))` to `~/.config/xonsh/rc.xsh`, Xonsh throws exceptions on initialization (examples with minimal traceback below). Happy to bring more detail if needed.

### Expected Behavior

Xonsh should initialize without exceptions when carapace is added to the run control.

### Steps To Reproduce

To try and track down what was happening, I added echo statements — `this is <whatever the file is>` — to the top of my Bash and Zsh run control files. My login shell is Zsh, but I do not use it — there's nothing in the run control file except the echo. All the examples below are run from iTerm2 profiles dedicated to the shell indicated.

### Examples with Bash, Zsh, and Elvish run control files available:

**From Xonsh:**
```python
this is bashrc
xonsh: For full traceback set: $XONSH_SHOW_TRACEBACK = True
  File "<string>", line 3
    get-env () { echo "${!1}"; } # this looks like Elvish, but I'm not getting a debug print statement from my `rc.elv`
               ^
SyntaxError: invalid syntax
syntax error in xonsh run control file '/Users/j0m0aam/.config/xonsh/rc.xsh': invalid syntax (<string>, line 3)

~
❯
```

**From Zsh:**
```python
~ % xonsh
zsh: command not found: xonsh 
~ % ~/.local/xonsh-env/bin/xonsh
xonsh: For full traceback set: $XONSH_SHOW_TRACEBACK = True
  File "<string>", line 3
    get-env () { echo "${(P)1}"; }
               ^
SyntaxError: invalid syntax
syntax error in xonsh run control file '<my home dir>/.config/xonsh/rc.xsh': invalid syntax (<string>, line 3)


~
❯
~ % ~/.local/xonsh-env/bin/xonsh --no-rc
~ @
```

### Examples without Bash, Zsh, and Elvish run control files available:

**From Xonsh:**
```
xonsh: For full traceback set: $XONSH_SHOW_TRACEBACK = True
  File "<string>", line 3
    get-env () { echo "${!1}"; }
               ^
SyntaxError: invalid syntax
syntax error in xonsh run control file '<my home dir>/.config/xonsh/rc.xsh': invalid syntax (<string>, line 3)
```

**From Zsh:***
```
 ~ % ~/.local/xonsh-env/bin/xonsh
xonsh: For full traceback set: $XONSH_SHOW_TRACEBACK = True
  File "<string>", line 3
    get-env () { echo "${(P)1}"; }
               ^
SyntaxError: invalid syntax
syntax error in xonsh run control file '<my home dir>/.config/xonsh/rc.xsh': invalid syntax (<string>, line 3)


~
❯
 ~ % ~/.local/xonsh-env/bin/xonsh --no-rc
 ~ @
```

### OS

- [X] Darwin
- [ ] Linux
- [ ] Termux
- [ ] Windows

### Shell

- [ ] Bash
- [ ] Cmd
- [ ] Elvish
- [ ] Fish
- [ ] Nushell
- [ ] Oil
- [ ] Powershell
- [X] Xonsh
- [ ] Zsh

**Comment by rsteube** [comment]
Looks like the [detection](https://github.com/carapace-sh/carapace/blob/master/pkg/ps/ps.go) isn't working. It's a simple check using the parent process names.

Do you still get the exceptions with:
```xonsh
# ~/.config/xonsh/rc.xsh
exec($(carapace _carapace xonsh))
```

**Comment by JeffMelton** [comment]
> Looks like the [detection](https://github.com/carapace-sh/carapace/blob/master/pkg/ps/ps.go) isn't working. It's a simple check using the parent process names.
> 
> 
> 
> Do you still get the exceptions with:
> 
> ```xonsh
> 
> # ~/.config/xonsh/rc.xsh
> 
> exec($(carapace _carapace xonsh))
> 
> ```

Nope, that works as expected. 👍  That's good enough for me. Thank you (and thanks for making carapace, et al.).

**Comment by JeffMelton** [comment]
A little more information that may be relevant: 

I used [xonsh-install](https://github.com/anki-code/xonsh-install) to set up my environment. It turns out that xonsh-install uses a Bash wrapper to launch Xonsh (see [anki-code/xonsh-install #4](https://github.com/anki-code/xonsh-install/issues/4) and other discussion it links back to). I don't use Xonsh as a login shell — rather, I set up a profile in my terminal emulator to launch the shell command directly. That uses `sh -c`, so by the time Xonsh launches we're at `$SHLVL` 2: `sh -c 'bash -i $HOME/.local/xonsh-env/bin/xonsh'`, more or less.

I don't know how carapace tries to detect the shell it's in, but it seems reasonable to suspect that the extra turtles in the stack might be getting in its way a bit.

**Comment by rsteube** [comment]
It just checks the parent process names one by one until it finds one that fits (see [here](https://github.com/carapace-sh/carapace/blob/master/pkg/ps/ps.go)).
LIkely endet up at the bash one and outputted the wrong script.

Nothing fancy, but usually does the job quite fine.
Might give it another thought if traversing further up is really a good idea, but as far as i remember i had a reason for doing it this way :thinking: .

**Comment by rsteube** [comment]
Probably looks similar to this for you as well?:
```sh
root@4f709ae6730a / @# ps -fo command
COMMAND
bash
bash
 \_ /bin/bash -i /root/.local/xonsh-env/xbin/xonsh
     \_ /root/.local/xonsh-env/bin/python /root/.local/xonsh-env/bin/xonsh
         \_ ps -fo command
```

**Comment by JeffMelton** [comment]
Yep, that's the stack of turtles this one puts together. 😆
